File: scala-server/app/com/bolour/eagerwords/common/domain/PieceProviderType.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.common.domain

object PieceProviderType {

  val RandomAsString = "Random"
  val CyclicAsString = "Cyclic"

  /**
    * Type of factory used to generate new pieces for use
    * in the game.
    */
  sealed abstract class PieceProviderType

  /**
    * Factory type to generate random pieces. The default.
    */
  object Random extends PieceProviderType {
    override def toString = RandomAsString
  }

  /**
    * Factory type to generate pieces by cycling through the alphabet.
    * Used for deterministic testing.
    */
  object Cyclic extends PieceProviderType {
    override def toString = CyclicAsString
  }

  def fromString(asString: String): PieceProviderType = {
    asString match {
      case RandomAsString => PieceProviderType.Random
      case CyclicAsString => PieceProviderType.Cyclic
      case _ => throw new RuntimeException(s"unrecognized piece provider type: ${asString}")
    }
  }
}

File: haskell-server/src/EagerWords/Common/Domain/PieceProviderType.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module EagerWords.Common.Domain.PieceProviderType(
  PieceProviderType(..)
  )
  where

import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)

-- | Different types of piece generators.
data PieceProviderType = Random | Cyclic
  deriving (Eq, Show, Generic, NFData)

instance FromJSON PieceProviderType
instance ToJSON PieceProviderType




